Compilation fails on `ReadWriteVector` with Kokkos complex datatypes

I am trying to build deal.II and Trilinos with clang 13.0.0 (aocc 3.2.0). Trilinos 15.0.0 was configured with `ENABLE_FLOAT=ON` and `ENABLE_COMPLEX=ON` and successfully compiled with said compiler.

Compiling deal.II with `DEAL_II_WITH_COMPLEX_VALUES=ON` yields errors of the following kind:
```
In file included from /dealii/source/lac/read_write_vector.cc:17:
/dealii/include/deal.II/lac/read_write_vector.templates.h:601:23: error: no viable overloaded '+='
            values[i] += new_values(i);
            ~~~~~~~~~ ^  ~~~~~~~~~~~~~
/dealii/include/deal.II/lac/read_write_vector.templates.h:825:5: note: in instantiation of function template specialization 'dealii::LinearAlgebra::ReadWriteVector<std::complex<float>>::import_elements<std::complex<float>>' requested here
    import_elements(trilinos_vec.trilinos_vector(),
    ^
/usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/complex:1132:7: note: candidate function not viable: no known conversion from 'const Kokkos::complex<float>' to 'float' for 1st argument
      operator+=(float __f)
      ^
/usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/complex:1174:9: note: candidate template ignored: could not match 'std::complex' against 'Kokkos::complex'
        operator+=(const complex<_Tp>& __z)
        ^
```

There are some additional errors of similar kind. The note I get consistently is `could not match 'std::complex' against 'Kokkos::complex'`.

Can you reproduce it?

--- a/include/deal.II/lac/read_write_vector.templates.h
+++ b/include/deal.II/lac/read_write_vector.templates.h
@@ -593,12 +593,12 @@ namespace LinearAlgebra
       {
         case VectorOperation::insert:
           for (size_type i = 0; i < size; ++i)
-            values[i] = new_values(i);
+            values[i] = Number(new_values(i));
           break;
 
         case VectorOperation::add:
           for (size_type i = 0; i < size; ++i)
-            values[i] += new_values(i);
+            values[i] += Number(new_values(i));
           break;
 
         case VectorOperation::min:
@@ -609,14 +609,14 @@ namespace LinearAlgebra
           for (size_type i = 0; i < size; ++i)
             {
               Assert(
-                std::imag(new_values(i)) == 0.,
+                std::imag(Number(new_values(i))) == 0.,
                 ExcMessage(
                   "VectorOperation::min is not defined if there is an imaginary part!)"));
               Assert(
                 std::imag(values[i]) == 0.,
                 ExcMessage(
                   "VectorOperation::min is not defined if there is an imaginary part!)"));
-              if (std::real(new_values(i)) - std::real(values[i]) < 0.0)
+              if (std::real(Number(new_values(i))) - std::real(values[i]) < 0.0)
                 values[i] = new_values(i);
             }
           break;
@@ -625,15 +625,15 @@ namespace LinearAlgebra
           for (size_type i = 0; i < size; ++i)
             {
               Assert(
-                std::imag(new_values(i)) == 0.,
+                std::imag(Number(new_values(i))) == 0.,
                 ExcMessage(
                   "VectorOperation::max is not defined if there is an imaginary part!)"));
               Assert(
                 std::imag(values[i]) == 0.,
                 ExcMessage(
                   "VectorOperation::max is not defined if there is an imaginary part!)"));
-              if (std::real(new_values(i)) - std::real(values[i]) > 0.0)
-                values[i] = new_values(i);
+              if (std::real(Number(new_values(i))) - std::real(values[i]) > 0.0)
+                values[i] = Number(new_values(i));
             }
           break;
 

--- a/include/deal.II/lac/trilinos_tpetra_vector.templates.h
+++ b/include/deal.II/lac/trilinos_tpetra_vector.templates.h
@@ -835,26 +835,34 @@ namespace LinearAlgebra
     bool
     Vector<Number, MemorySpace>::is_non_negative() const
     {
-      // get a representation of the vector and
-      // loop over all the elements
-      Teuchos::ArrayRCP<const Number> data       = vector->getData();
-      const size_type                 n_elements = vector->getLocalLength();
-      unsigned int                    flag       = 0;
-      for (size_type i = 0; i < n_elements; ++i)
+      if constexpr (!std::is_same_v<Number, std::complex<double>> &&
+                    !std::is_same_v<Number, std::complex<float>>)
         {
-          if (data[i] < Number(0))
+          // get a representation of the vector and
+          // loop over all the elements
+          Teuchos::ArrayRCP<const Number> data       = vector->getData();
+          const size_type                 n_elements = vector->getLocalLength();
+          unsigned int                    flag       = 0;
+          for (size_type i = 0; i < n_elements; ++i)
             {
-              flag = 1;
-              break;
+              if (data[i] < Number(0))
+                {
+                  flag = 1;
+                  break;
+                }
             }
-        }
 
-      // Check that the vector is non-negative on _all_ processors.
-      unsigned int num_negative =
-        Utilities::MPI::sum(flag,
-                            Utilities::Trilinos::teuchos_comm_to_mpi_comm(
-                              vector->getMap()->getComm()));
-      return num_negative == 0;
+          // Check that the vector is non-negative on _all_ processors.
+          unsigned int num_negative =
+            Utilities::MPI::sum(flag,
+                                Utilities::Trilinos::teuchos_comm_to_mpi_comm(
+                                  vector->getMap()->getComm()));
+          return num_negative == 0;
+        }
+      Assert(false,
+             ExcMessage("You can't ask a complex value "
+                        "whether it is non-negative."));
+      return true;
     }
 
 

